hip/base/scoped_device_id.hip.cpp

```diff
@@ -41,59 +41,57 @@ OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
 
 #include <ginkgo/core/base/exception_helpers.hpp>
+#include <ginkgo/core/base/scoped_device_id.hpp>
 
 
 namespace gko {
-namespace hip {
-
-
-/**
- * This class defines a device guard for the hip functions and the hip module.
- * The guard is used to make sure that the device code is run on the correct
- * hip device, when run with multiple devices. The class records the current
- * device id and uses `hipSetDevice` to set the device id to the one being
- * passed in. After the scope has been exited, the destructor sets the device_id
- * back to the one before entering the scope.
- */
-class device_guard {
-public:
-    device_guard(int device_id) : original_device_id{}, need_reset{}
-    {
-        GKO_ASSERT_NO_HIP_ERRORS(hipGetDevice(&original_device_id));
-        if (original_device_id != device_id) {
-            GKO_ASSERT_NO_HIP_ERRORS(hipSetDevice(device_id));
-            need_reset = true;
-        }
-    }
+namespace detail {
 
-    device_guard(device_guard& other) = delete;
 
-    device_guard& operator=(const device_guard& other) = delete;
-
-    device_guard(device_guard&& other) = delete;
+hip_scoped_device_id::hip_scoped_device_id(int device_id)
```

Comment:
I guess this is not a interface break, because this is not a user-facing header file. But this would break ABI though.